### ISSUE TITLE
ServerlessError: An error occurred: ServerlessDeploymentBucket - API: s3:setBucketAccelerateConfiguration Access Denied.

### ISSUE THREAD

**Issue by dperszyk** [serverless/serverless#4625]
# This is a Bug Report
## Description

For bug reports:
* What went wrong?
The deploy failed with error `ServerlessError: An error occurred: ServerlessDeploymentBucket - API: s3:setBucketAccelerateConfiguration Access Denied.`. We've been using serverless with the same IAM user for the past few months, and this is the first time it's happened.
* What did you expect should have happened?
I'm guessing this has something to do with the new feature added for bucketAcceleration. I would expect that the addition of that feature would not break current deploys -- I don't want bucket acceleration, so I shouldn't have to give that permission to my IAM user.
* What was the config you used?
* What stacktrace or error message from your provider did you see?
```ServerlessError: An error occurred: ServerlessDeploymentBucket - API: s3:setBucketAccelerateConfiguration Access Denied.
    at provider.request.then (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/lib/monitorStack.js:112:33)
From previous event:
    at AwsDeploy.monitorStack (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/lib/monitorStack.js:26:12)
    at provider.request.then (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/lib/updateStack.js:84:30)
From previous event:
    at AwsDeploy.update (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/lib/updateStack.js:84:8)
From previous event:
    at AwsDeploy.BbPromise.bind.then (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/lib/updateStack.js:101:12)
From previous event:
    at AwsDeploy.updateStack (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/lib/updateStack.js:95:8)
From previous event:
    at AwsDeploy.BbPromise.bind.then (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/deploy/index.js:135:39)
From previous event:
    at Object.AwsDeploy.hooks.aws:deploy:deploy:updateStack [as hook] (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/deploy/index.js:131:10)
    at BbPromise.reduce (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/classes/PluginManager.js:368:55)
From previous event:
    at PluginManager.invoke (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/classes/PluginManager.js:368:22)
    at PluginManager.spawn (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/classes/PluginManager.js:386:17)
    at AwsDeploy.BbPromise.bind.then (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/deploy/index.js:101:48)
From previous event:
    at Object.AwsDeploy.hooks.deploy:deploy [as hook] (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/plugins/aws/deploy/index.js:97:10)
    at BbPromise.reduce (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/classes/PluginManager.js:368:55)
From previous event:
    at PluginManager.invoke (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/classes/PluginManager.js:368:22)
    at PluginManager.run (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/classes/PluginManager.js:399:17)
    at variables.populateService.then (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/Serverless.js:102:33)
    at tryOnImmediate (timers.js:543:15)
    at processImmediate [as _immediateCallback] (timers.js:523:5)
From previous event:
    at Serverless.run (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/lib/Serverless.js:89:74)
    at serverless.init.then (/opt/circleci/nodejs/v6.1.0/lib/node_modules/serverless/bin/serverless:42:50)
```
For feature proposals:
* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.
* If there is additional config how would it look

Similar or dependent issues:
* #12345

## Additional Data

* ***Serverless Framework Version you're using***: 1.25.0
* ***Operating System***:
* ***Stack Trace***:
* ***Provider Error messages***:

**Comment by horike37** [comment]
Hi @dperszyk, thank you for repoting :+1: :smile: 

>I'm guessing this has something to do with the new feature added for bucketAcceleration.

I think so.
Which commands did you run when facing the error? `sls deploy`?
Additionally, could you tell us policies list that the IAM user is attached so that we can reproduce it on our end?

**Comment by dperszyk** [comment]
Hi @horike37 !

Thanks for responding so quickly. 
I ran `sls deploy --stage MYENV --aws-profile MYAWSPROFILE`.

Here's the policy for the IAM user, but with the resource names changed: 

```        Statement=[
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("s3", "GetObject"),
                        Action("s3", "PutObject"),
                        Action("s3", "ListBucket"),
                        Action("s3", "DeleteObject"),
                        Action("s3", "DeleteBucket"),
                        Action("s3", "ListBucketVersions"),
                        Action("s3", "GetBucketNotification"),
                        Action("s3", "PutBucketNotification")
                    ],
                    Resource=[
                        "arn:aws:s3:::mylambdas*serverlessdeploymentbucket*"
                    ]
                ),
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("s3", "CreateBucket"),
                    ],
                    Resource=[
                        "arn:aws:s3:::*"
                    ]
                ),
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("iam", "CreateRole"),
                        Action("iam", "PassRole"),
                        Action("iam", "AttachRolePolicy"),
                        Action("iam", "CreatePolicy"),
                        Action("apigateway", "GET"),
                        Action("apigateway", "POST"),
                        Action("apigateway", "PUT"),
                        Action("apigateway", "DELETE")
                    ],
                    Resource=["*"]
                ),
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("cloudformation", "List*"), 
                        Action("cloudformation", "Get*"), 
                        Action("cloudformation", "PreviewStackUpdate"),
                        Action("cloudformation", "ValidateTemplate")
                    ],
                    Resource=[
                        "*"
                    ]
                ), 
                Statement(
                    Effect=Allow, 
                    Action=[
                        Action("cloudformation", "CreateStack"),
                        Action("cloudformation", "CreateUploadBucket"),
                        Action("cloudformation", "DeleteStack"),
                        Action("cloudformation", "DescribeStackEvents"),
                        Action("cloudformation", "DescribeStackResource"),
                        Action("cloudformation", "DescribeStackResources"),
                        Action("cloudformation", "UpdateStack"),
                        Action("cloudformation", "DescribeStacks")
                    ], 
                    Resource=[
                        "arn:aws:cloudformation:*:*:stack/mylambdas-*/*"
                    ]
                ),
                Statement(
                    Effect=Allow, 
                    Action=[
                        Action("lambda", "*")
                    ], 
                    Resource=[
                        "*"
                    ]
                ),
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("iam", "*")
                    ],
                    Resource=["arn:aws:iam::*:role/mylambdas-*-*-myLambdaRole"]
                ),
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("sqs", "*")
                    ],
                    Resource=[
                        "arn:aws:sqs:*:*:mylambdas-*-*"
                    ]
                ),
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("kinesis", "*")
                    ],
                    Resource=["arn:aws:kinesis:*:*:stream/mylambdas-*-*"]
                ),
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("events", "*")
                    ],
                    Resource=["*"]
                ),
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("dynamodb", "*")
                    ],
                    Resource=[
                        "*"
                    ]
                ),
                Statement(
                    Effect=Allow,
                    Action=[
                        Action("cloudwatch", "GetMetricsStatistics")
                    ],
                    Resource=["*"]
                ), 
                Statement(
                    Effect=Allow, 
                    Action=[
                        Action("logs", "CreateLogGroup"),
                        Action("logs", "DeleteLogGroup"), 
                        Action("logs", "CreateLogStream"), 
                        Action("logs", "PutLogEvents"),
                        Action("logs", "DescribeLogStreams"),
                        Action("logs", "DescribeLogGroups"),
                        Action("logs", "FilterLogEvents")
                    ], 
                    Resource=[
                        "*"
                    ]
                )
            ]
```

Let me know if you need anything else. Thanks again!

**Comment by horike37** [comment]
Thank you for the detail @dperszyk :+1: 

I took a deep debugging.
The root cause is that `AccelerateConfiguration` is needed within generated CloudFormation template even if you deploy without `--aws-s3-accelerate`.
Here is `cloudformation-template-create-stack.json` of this case. As a result, CF needs `PutAccelerateConfiguration` policy whether with the config or not.

```
"Resources": {
    "ServerlessDeploymentBucket": {
      "Type": "AWS::S3::Bucket",
      "Properties": {
        "AccelerateConfiguration": {
          "AccelerationStatus": "Suspended"
        }
      }
    }
  }
```

Imo, That's a breaking change for users who set up minimum IAM policy.
It would be good to remove `AccelerateConfiguration` statement if you deploy without the config if it is possible.

cc: @alexcasalboni

**Comment by alexcasalboni** [comment]
@horike37 I think you are right. You can find the related code [here](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/package/lib/generateCoreTemplate.js#L52).

The `Suspended` value was intended to turn the feature off as soon as you stop using `--aws-s3-accelerate`. This is, in fact, problematic without `s3:PutAccelerateConfiguration` permissions.

As a quick fix, we could simply disable the `Suspended` value and only use `Enabled`. Please note that if we do NOT force the `Suspended` value, we won't disable Transfer Acceleration when you deploy without `--aws-s3-accelerate` (we discussed this case in #4293). That's objectively less annoying than this breaking change.

What do you think?

**Comment by horike37** [comment]
@alexcasalboni 

>Please note that if we do NOT force the Suspended value, we won't disable Transfer Acceleration >when you deploy without --aws-s3-accelerate (we discussed this case in #4293). 

Ah, I see. TBH, I can't think of any best solution 😅 😅 😅 
It would be good to collect some feedback from someone else of the community.

cc: @serverless/framework-maintainers

**Comment by HyperBrain** [comment]
Hi @horike37 , @alexcasalboni . I would also go without the automatic disable. Having clean minimal permissions is rated higher for me.
What about having a separate switch to disable the acceleration on demand again on the deployment bucket? So --aws-s3-accelerate would turn it on and `--no-aws-s3-accelerate` would be the symmetric part to turn it off again explicitly.

**Comment by alexcasalboni** [comment]
@HyperBrain @horike37 @dperszyk  it should be fixed in #4636 :)

Please have a look and let me know what you think.

**Comment by HyperBrain** [comment]
@alexcasalboni That was fast 😄 . We will check and review the PR soon.
